src/schedule/ll_schedule.c

```diff
@@ -128,20 +125,30 @@ static void schedule_ll_clients_reschedule(struct ll_schedule_data *sch)
 static void schedule_ll_tasks_run(void *data)
 {
 	struct ll_schedule_data *sch = data;
+	uint32_t num_clients;
 	uint32_t flags;
 
 	domain_disable(sch->domain, cpu_get_id());
 
 	irq_local_disable(flags);
 
+	spin_lock(sch->domain->lock);
+
+	/* clear domain only if all clients are done */
+	num_clients = atomic_sub(&sch->domain->num_clients, 1);
+	if (!num_clients)
+		domain_clear(sch->domain);
+
+	spin_unlock(sch->domain->lock);
+
```

Comment:
Do we really need lock if sub is atomic? Are you worried about domain_clear being executed in parallel? If subs are atomic there should be only 1 client that gets num_clients to 0 and calls domain clear, unless it's called also from some other place.

Reply:
What about removing the last client but in the meanwhile a new client appears? Then you'd clear the domain with someone who uses it alive, which I think is a _really_ bad idea.

Reply:
Ah ok, so you need it for that case. Your are right then.

Reply:
@paulstelian97 @jajanusz Yes, this is the exact problem. Thousands of multicore cycles showed that sometimes we could have race here.